File: app/schemas/roll_v2.py
"""Roll v2 bootstrap and rate response schemas.

Contract-only frozen DTOs for issue #2716. No synchronous provider calls and
no projection work live here; the v2 bootstrap derives canonical identity
from stored mappings only (unavailable identity yields null canonical id and
null canonical-series stats per #1401, never a thread-title substitution).
"""

from datetime import datetime
from enum import Enum
from typing import ClassVar, Literal

from pydantic import BaseModel, Field, field_validator, model_validator

from app.schemas.roll import RollRecoveryInfo
from app.schemas.session import ActiveThreadInfo, SessionBandwidthState, SessionMode
from app.schemas.thread import ThreadResponse


class IdentityState(str, Enum):
    """Derived identity states for rollable items."""

    CONFIRMED = "confirmed"
    CANDIDATE = "candidate"
    UNRESOLVED = "unresolved"
    AMBIGUOUS = "ambiguous"
    CONFLICTING = "conflicting"


class RouteKind(str, Enum):
    """Route kind for rollable items."""

    GROUP = "group"


class ProgressScope(str, Enum):
    """Progress scope for reader context."""

    CANONICAL_SERIES_RUN = "canonical_series_run"
    THREAD = "thread"


class RollableThread(BaseModel):
    """Thread information for rollable items."""

    id: int
    title: str
    format: str
    last_activity_at: str | None = None


class RollableIssue(BaseModel):
    """Issue information for rollable items."""

    id: int
    number: str
    canonical_series_title: str | None = None
    cover_url: str | None = None

    @field_validator("cover_url")
    @classmethod
    def cover_must_be_same_origin(cls, value: str | None) -> str | None:
        """Enforce the frozen cover contract: same-origin optimized URLs only.

        Args:
            value: Candidate cover URL.

        Returns:
            The value unchanged when null or same-origin.

        Raises:
            ValueError: If a non-null cover is not a same-origin
                ``/api/v1/images/optimize?...`` URL.
        """
        if value is None:
            return None
        if value.startswith("/api/v1/images/optimize"):
            return value
        raise ValueError("Roll v2 covers must be same-origin /api/v1/images/optimize URLs")


class RollableIdentity(BaseModel):
    """Identity information for rollable items."""

    source: Literal["comicvine", "unavailable"]
    canonical_series_id: str | None = None
    state: IdentityState
    series_mapping_state: IdentityState


class RollableReader(BaseModel):
    """Reader context for rollable items."""

    latest_rating: float | None = None
    average_rating: float | None = None
    rating_count: int | None = None
    read_count: int | None = None
    issue_count: int | None = None
    progress_scope: ProgressScope


class RollableRoute(BaseModel):
    """Route information for rollable items."""

    kind: RouteKind = Field(default=RouteKind.GROUP)
    name: str


class RollableItem(BaseModel):
    """A single rollable candidate thread with issue and context."""

    thread: RollableThread
    issue: RollableIssue  # Required/non-null as per spec
    identity: RollableIdentity
    reader: RollableReader
    routes: list[RollableRoute] = Field(default_factory=list, max_length=3)
    overflow_routes_count: int = Field(default=0, ge=0)


class RollLastRead(BaseModel):
    """Session-scoped last read information."""

    issue_id: int | None = None
    issue_number: str | None = None
    thread_id: int | None = None
    thread_title: str | None = None
    read_at: datetime | None = None


class RollV2BootstrapResponse(BaseModel):
    """V2 bootstrap response superset with rollable and last_read fields.

    Replaces the v1 roll_pool with rollable[] and adds nullable last_read.
    Maintains all v1 session/recovery/partition fields with unchanged semantics.
    """

    session_id: int
    user_id: int
    summary_limit: ClassVar[int] = 20

    current_die: int
    manual_die: int | None
    pending_thread_id: int | None
    last_rolled_result: int | None
    session_mode: SessionMode
    active_thread: ActiveThreadInfo | None
    roll_recovery: RollRecoveryInfo | None = None
    bandwidth: SessionBandwidthState
    rollable: list[RollableItem]  # Replaces roll_pool
    last_read: RollLastRead | None  # New nullable session-scoped field
    snoozed_threads: list[RollableThread] = Field(default_factory=list)
    snoozed_count: int
    skipped_thread_ids: list[int] = []
    skipped_threads: list[RollableThread] = []
    blocked_count: int
    blocked_threads: list[RollableThread] = []
    stale_thread_count: int
    stale_thread: RollableThread | None
    timezone: str | None = None

    @model_validator(mode="before")
    @classmethod
    def bound_summary_lists(cls, data: object) -> object:
        """Keep summary collections bounded even when stored session IDs grow."""
        if not isinstance(data, dict):
            return data

        for field_name in ("snoozed_threads", "skipped_threads", "blocked_threads"):
            values = data.get(field_name)
            if isinstance(values, list):
                data[field_name] = values[: cls.summary_limit]
        return data


class RollReconciliation(BaseModel):
    """Roll reconciliation information for rate response."""

    last_read: RollLastRead


class RateResponse(ThreadResponse):
    """Rate response extending ThreadResponse with roll reconciliation.

    Subclasses ThreadResponse so every current top-level field is preserved
    exactly, and adds ``roll_reconciliation.last_read`` for the v2 contract.
    Declared as the FastAPI ``response_model`` so the extra field is not
    filtered out.
    """

    roll_reconciliation: RollReconciliation | None = None

File: frontend/src/types/rollV2.ts
import type { SessionMode, SessionThread } from './index'
import type { RollRecoveryInfo } from './rollBootstrap'

export type { RollBootstrapResponse, RollBootstrapThread } from './rollBootstrap'

/** Frozen v2 identity derived-state enum. */
export type RollV2IdentityState =
  | 'confirmed'
  | 'candidate'
  | 'unresolved'
  | 'ambiguous'
  | 'conflicting'

/** Frozen v2 route kind. No subtype is inferred from the group name. */
export type RollV2RouteKind = 'group'

/** Frozen v2 progress scope. Canonical run length is the denominator. */
export type RollV2ProgressScope = 'canonical_series_run' | 'thread'

/** Frozen v2 identity source. Unavailable means null canonical id/stats. */
export type RollV2IdentitySource = 'comicvine' | 'unavailable'

/** Thread row inside a v2 rollable item. */
export interface RollV2Thread {
  id: number
  title: string
  format: string
  last_activity_at?: string | null
}

/** Required next-unread issue row. V2 never returns `issue: null`. */
export interface RollV2Issue {
  id: number
  number: string
  canonical_series_title?: string | null
  /** Same-origin `/api/v1/images/optimize?...` URL only, else null. */
  cover_url?: string | null
}

/** Canonical identity row reusing #1401 semantics. */
export interface RollV2Identity {
  source: RollV2IdentitySource
  canonical_series_id: string | null
  state: RollV2IdentityState
  series_mapping_state: RollV2IdentityState
}

/** Reader context row with explicit nullability. */
export interface RollV2Reader {
  latest_rating: number | null
  average_rating: number | null
  rating_count: number | null
  read_count: number | null
  issue_count: number | null
  progress_scope: RollV2ProgressScope
}

/** One route. At most 3 per item; extras count toward overflow. */
export interface RollV2Route {
  kind: RollV2RouteKind
  name: string
}

/** One rollable candidate row. */
export interface RollV2Item {
  thread: RollV2Thread
  issue: RollV2Issue
  identity: RollV2Identity
  reader: RollV2Reader
  routes: RollV2Route[]
  overflow_routes_count: number
}

/** Nullable session-scoped last-read object shared with rate responses. */
export interface RollV2LastRead {
  issue_id?: number | null
  issue_number?: string | null
  thread_id?: number | null
  thread_title?: string | null
  read_at?: string | null
}

/** Session bandwidth state carried by bootstrap responses. */
export interface RollV2BandwidthState {
  predicted_bandwidth: string | null
  active_bandwidth: string | null
  confidence: number | null
  source: string | null
  mode_version: string | null
}

/**
 * V2 bootstrap superset. Every v1 session/recovery/partition field keeps its
 * semantics; `roll_pool` is replaced by `rollable` and `last_read` is added.
 */
export interface RollV2BootstrapResponse {
  session_id: number
  user_id: number
  current_die: number
  manual_die: number | null
  pending_thread_id: number | null
  last_rolled_result: number | null
  session_mode: SessionMode
  active_thread: SessionThread | null
  roll_recovery?: RollRecoveryInfo | null
  bandwidth: RollV2BandwidthState
  rollable: RollV2Item[]
  last_read: RollV2LastRead | null
  snoozed_threads: RollV2Thread[]
  snoozed_count: number
  skipped_thread_ids: number[]
  skipped_threads: RollV2Thread[]
  blocked_count: number
  blocked_threads: RollV2Thread[]
  stale_thread_count: number
  stale_thread: RollV2Thread | null
  timezone?: string | null
}

/** Additive rate response: ThreadResponse fields plus roll reconciliation. */
export interface RollV2RateResponse {
  id: number
  title: string
  format: string
  issues_remaining: number
  queue_position: number
  status: string
  last_rating: number | null
  last_activity_at: string | null
  notes: string | null
  is_test: boolean
  is_blocked: boolean
  blocking_reasons: string[]
  created_at: string
  total_issues: number | null
  reading_progress: string | null
  next_unread_issue_id: number | null
  next_unread_issue_number: string | null
  roll_reconciliation: { last_read: RollV2LastRead } | null
}
